File: recipe-view/recipe-view/Models/Recipe.swift
import Foundation

struct Recipe: Decodable {
    let id: String
    let title: String
    let ingredients: [Ingredient]
    let instructions: [InstructionWithTime]
    let cookTimeAmount: Double?
    let cookTimeUnit: String?
	let servingsAmount: Double?
	let servingsUnit: String?
	let difficulty: String?
    let createdAt: String
    let updatedAt: String
    let userId: String?
    let url: String
    let urlHash: String	
    
    enum CodingKeys: String, CodingKey {
        case id
        case title
        case ingredients
        case instructions
        case cookTimeAmount = "cook_time_amount"
        case cookTimeUnit = "cook_time_unit"
		case servingsAmount = "servings_amount"
		case servingsUnit = "servings_unit"
        case difficulty
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case userId = "user_id"
        case url
        case urlHash = "url_hash"
    }
}

struct Ingredient: Decodable, Identifiable {
    let item: String
    let amount: Double
    let unit: String
    let category: Category
    let preparation: String?
    let isOptional: Bool
    let notes: String?
    
    var id: String { item }
    
    enum Category: String, Decodable {
        case protein = "protein"
        case grains = "grains"
        case produce = "produce"
        case oils = "oils"
        case condiments = "condiments"
        case spices = "spices"
        case sauces = "sauces"
        case dairy = "dairy"
        case herbs = "herbs"
        case other = "other"
        
        var displayName: String {
            rawValue.capitalized
        }
        
        // Static mapping of emojis
        static let categoryEmojis: [Category: String] = [
            .protein: "🥩",
            .grains: "🌾",
            .produce: "🥬",
            .oils: "🫗",
            .condiments: "🧂",
            .spices: "🌶️",
            .sauces: "🥫",
            .dairy: "🥛",
            .herbs: "🌿",
            .other: "📦"
        ]
        
        var emoji: String {
            Category.categoryEmojis[self] ?? "📦"
        }
    }
    
    enum CodingKeys: String, CodingKey {
        case item
        case amount
        case unit
        case category
        case preparation
        case isOptional = "is_optional"
        case notes
    }
}

struct InstructionWithTime: Decodable, Identifiable {
    let id: String
    let recipeId: String
    let step: Int
    let description: String
    let timeAmount: Double?
    let timeUnit: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case recipeId = "recipe_id"
        case step = "step_number"
        case description
        case timeAmount = "time_amount"
        case timeUnit = "time_unit"
    }
}

struct Time: Decodable {
    let amount: Double
    let unit: String
}

File: python-backend/src/schemas/recipe.py
from typing import List, Literal, Optional, Union
from pydantic import BaseModel, Field


class Ingredient(BaseModel):
    """
    Represents an ingredient in a recipe.
    """

    item: str = Field(..., description="The name of the ingredient")
    amount: Union[float, int] = Field(..., description="The quantity of the ingredient")
    unit: str = Field(..., description="Unit of measurement")

    category: Literal[
        "protein",
        "grains",
        "produce",
        "oils",
        "condiments",
        "spices",
        "sauces",
        "dairy",
        "herbs",
        "other",
    ] = Field(default="other", description="High-level category for grouping")

    preparation: Optional[str] = Field(
        None, description="Preparation instructions (e.g., crushed, julienned)"
    )

    is_optional: bool = Field(
        default=False, description="Whether ingredient is optional"
    )

    notes: Optional[str] = Field(
        None, description="Additional notes about the ingredient"
    )


class Time(BaseModel):
    """
    Represents the time needed in a recipe.
    # TODO: This will eventually map to a time UI component in the frontend
    """

    amount: Union[float, int] = Field(..., description="The quantity of time needed")
    unit: str = Field(..., description="The unit of measurement for the time")


class Instruction(BaseModel):
    """
    Represents a cooking instruction in a recipe.
    """

    step: int = Field(..., description="The step number in the instruction sequence")
    description: str = Field(..., description="The description of the instruction")
    time: Optional[Time] = Field(
        None, description="The time specified in the instruction"
    )


class ServingSize(BaseModel):
    """
    Represents the serving size of a recipe.
    """

    amount: Union[float, int] = Field(
        ..., description="The quantity of the serving size"
    )
    unit: str = Field(..., description="The unit of measurement for the serving size")


class Recipe(BaseModel):
    """Represents a simple recipe object."""

    title: str = Field(..., description="The name of the recipe")
    servings: ServingSize = Field(
        ..., description="The number of servings the recipe makes"
    )
    cookTime: Optional[Time] = Field(None, description="The cook time for the recipe")
    difficulty: Literal["novice", "home_cook", "professional_chef", "other"] = Field(
        ..., description="The difficulty of the recipe"
    )
    ingredients: List[Ingredient] = Field(
        ..., description="The ingredients needed for the recipe"
    )
    instructions: List[Instruction] = Field(
        ..., description="The instructions for the recipe"
    )
